Sources/SyncDataProviders/Credentials/internal/CredentialsResponseHandler.swift

```diff
@@ -0,0 +1,199 @@
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import BrowserServicesKit
+import DDGSync
+import Foundation
+import GRDB
+
+final class CredentialsResponseHandler {
+    let clientTimestamp: Date
+    let received: [SyncableCredentialsAdapter]
+    let secureVault: any AutofillSecureVault
+    let database: Database
+    let shouldDeduplicateEntities: Bool
+
+    let allReceivedIDs: Set<String>
+    private var credentialsByUUID: [String: SecureVaultModels.SyncableCredentials] = [:]
+
+    private let decrypt: (String) throws -> String
+
+    init(received: [Syncable], clientTimestamp: Date, secureVault: any AutofillSecureVault, database: Database, crypter: Crypting, deduplicateEntities: Bool) throws {
+        self.clientTimestamp = clientTimestamp
+        self.received = received.map(SyncableCredentialsAdapter.init)
+        self.secureVault = secureVault
+        self.database = database
+        self.shouldDeduplicateEntities = deduplicateEntities
+
+        let secretKey = try crypter.fetchSecretKey()
+        self.decrypt = { try crypter.base64DecodeAndDecrypt($0, using: secretKey) }
+
+        var allUUIDs: Set<String> = []
+
+        self.received.forEach { syncable in
+            guard let uuid = syncable.uuid else {
+                return
+            }
+            allUUIDs.insert(uuid)
+        }
+
+        self.allReceivedIDs = allUUIDs
+
+        credentialsByUUID = try secureVault.syncableCredentialsForSyncIds(allUUIDs, in: database).reduce(into: .init(), { $0[$1.metadata.uuid] = $1 })
+    }
+
+    func processReceivedCredentials() throws {
+        if received.isEmpty {
+            return
+        }
+
+        let encryptionKey = try secureVault.getEncryptionKey()
+        let hashingSalt = try secureVault.getHashingSalt()
+
+        for syncable in received {
+            try processEntity(with: syncable, secureVaultEncryptionKey: encryptionKey, secureVaultHashingSalt: hashingSalt)
+        }
+    }
+
+    // MARK: - Private
+
+    private func processEntity(with syncable: SyncableCredentialsAdapter, secureVaultEncryptionKey: Data, secureVaultHashingSalt: Data?) throws {
+        guard let syncableUUID = syncable.uuid else {
+            throw SyncError.accountAlreadyExists // todo
+        }
+
+        if shouldDeduplicateEntities, var deduplicatedEntity = try deduplicatedCredentials(with: syncable, secureVaultEncryptionKey: secureVaultEncryptionKey) {
+            let oldUUID = deduplicatedEntity.metadata.uuid
+            deduplicatedEntity.account?.title = try syncable.encryptedTitle.flatMap(decrypt)
```

Comment:
Based on the spec, we should be looking at last modified when deduplicating these fields

Reply:
Since we aren't supposed to compare against server timestamp, I'm going to bring this back to Asana.